File: app/scripts/example-dates.mjs
/**
 * The example instance's records, as old today as they were written to be.
 *
 * The whole of the rule, and why it is a rule rather than a fresher fixture,
 * is stated once in `tools/scripts/example_dates.py`. This is the same rule in
 * the language the cockpit's build speaks: the demonstration's own records are
 * compiled into the bundle by `./example-instance.mjs`, and no Python runs
 * where that happens.
 *
 * The two are pinned to each other by `tools/tests/fixtures/example-dates.json`
 * -- the same anchor, the same worked cases, read from both sides (D-14). A
 * shift the build applied and the second-instance build did not would put the
 * cockpit's own copy of a session on a different evening from the showcase's,
 * which is the one thing a demonstration of a single instance may not do.
 *
 * `CONVENER_EXAMPLE_TODAY` is read here for the same single caller:
 * `tools/scripts/render_readme_shots.py` sets it for the whole build so that
 * the pictures `README.md` shows stay byte-identical between two runs on
 * different days.
 */

/** The day `examples/the-example-collective/instance/data/` is written for.
 *  Mirrors `example_dates.ANCHOR`. */
export const ANCHOR = '2026-09-03';

/** The one environment variable that replaces the clock here. */
export const TODAY_ENV = 'CONVENER_EXAMPLE_TODAY';

/** An ISO day, bounded on both sides so a longer run of digits is not
 *  half-matched. Mirrors `example_dates.ISO_DAY`. */
const ISO_DAY = /(?<!\d)(\d{4})-(\d{2})-(\d{2})(?!\d)/g;

const MS_PER_DAY = 86_400_000;

/** Midnight UTC of an ISO day, or `null` when the calendar has no such day.
 *  The round trip is what refuses `2026-02-30`, which `Date.parse` accepts
 *  and rolls forward to 2 March -- `app/src/state/sla.ts::parsedDay` refuses
 *  it the same way and for the same reason. */
function parse(day) {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(day)) return null;
  const epoch = Date.parse(`${day}T00:00:00Z`);
  if (Number.isNaN(epoch)) return null;
  return new Date(epoch).toISOString().slice(0, 10) === day ? epoch : null;
}

function isoOf(epoch) {
  return new Date(epoch).toISOString().slice(0, 10);
}

/**
 * The day the example's records are read against.
 *
 * Throws on a value that is set and unreadable, the way the Python side does:
 * a renderer that meant to pin the clock and misspelled the day would
 * otherwise photograph a moving fixture and report success (D-25).
 */
export function today(env = process.env) {
  const pinned = env[TODAY_ENV] ?? '';
  if (!pinned) return new Date().toISOString().slice(0, 10);
  if (parse(pinned) === null) {
    throw new Error(
      `${TODAY_ENV} is ${JSON.stringify(pinned)}, which is not a YYYY-MM-DD day -- ` +
        "it is read to pin the example instance's own records to one moment, and a " +
        'value nothing can parse would leave them moving'
    );
  }
  return pinned;
}

/** Whole days -- always a multiple of seven -- from the anchor to the week
 *  `day` falls in. `Math.floor` and not a truncation: the shift is negative
 *  for a day before the anchor, which is the ordinary case for a pinned
 *  renderer, and truncating would round it the wrong way. */
export function shiftDays(day) {
  const now = parse(day);
  if (now === null) throw new Error(`${JSON.stringify(day)} is not a YYYY-MM-DD day`);
  const anchor = parse(ANCHOR);
  return Math.floor((now - anchor) / MS_PER_DAY / 7) * 7;
}

/** `text` with every ISO day in it moved into the week of `day`. A date the
 *  calendar does not have is left exactly as it was found. */
export function shifted(text, day) {
  const days = shiftDays(day);
  if (days === 0) return text;
  return text.replace(ISO_DAY, (whole) => {
    const found = parse(whole);
    return found === null ? whole : isoOf(found + days * MS_PER_DAY);
  });
}

File: tools/scripts/example_dates.py
"""The example instance's records, as old today as they were written to be.

`examples/the-example-collective/instance/data/` is a fixture with dates in
it, and a fixture with dates in it goes stale on its own. Read on the day
they were written, the records show a board with a vote open, a session five
weeks out and a wrap-up nobody is late on; read eleven months later they show
the same records with every deadline behind them, and the demonstration's
inbox prints "Forum summary is 281 days overdue" about a talk nobody gave.
That is not a defect in the inbox -- the inbox is right -- and it is not a
defect in the reader either: `app/src/data/demo.ts` reads the example rather
than inventing, deliberately. It is a defect in the days.

What this does, and the whole of it
-----------------------------------
Every ISO day in the text moves by the same whole number of weeks, so that
the fixture's own present lands on the week the demonstration is being
watched in.

    shift = floor((today - ANCHOR) / 7) * 7 days

**Weeks rather than days**, because a session is on a Thursday evening and
`agenda.py` composes the calendar entry from the day it finds: shifting by
whole weeks keeps every weekday it was written with, and shifting by days
would move a monthly Thursday series onto a Tuesday. What follows from that
is that the fixture's own present is between zero and six days behind the
real one, never ahead, so a record written as "due in seven days" is never
read as overdue.

**Text rather than records.** The rule is one regular expression over the
file, not a walk of the schema, and that is the point: a field added to the
model next year carries its date through this without anybody remembering to
name it here, and nothing here can reformat a file or drop the comment at the
top of it. The comment moves too, which is right rather than a side effect --
the shifted document says which day it is written for, and it is a different
day from the one the tracked file is written for.

Two implementations, one fixture
--------------------------------
`app/scripts/example-dates.mjs` is the same rule in the language the cockpit's
build speaks, because the demonstration's own records are compiled into that
bundle (`app/scripts/example-instance.mjs`) and no Python runs where that
happens. `tools/tests/fixtures/example-dates.json` is what stops the two
drifting: both sides read the same anchor and the same worked cases out of it
and must answer identically, the way `edition-prefix.json` already binds the
two readers of the declaration (D-14).

The day it is read against, and the one place it is ever overridden
-------------------------------------------------------------------
The real day, from the clock, everywhere but one: `CONVENER_EXAMPLE_TODAY`.
`tools/scripts/render_readme_shots.py` sets it to the day it also pins the
browser's clock to, so that the pictures in `README.md` are one moment in
the example instance's life and stay byte-identical between two runs on
different days. Nothing else sets it, and nothing about a real instance passes
through here at all: a repository's own `instance/data/` is never read by this
module, only the example the product ships.
"""

from __future__ import annotations

import datetime as dt
import os
import re
from collections.abc import Mapping
from typing import Final

#: The day `examples/the-example-collective/instance/data/` is written for.
#: Every date in those files is placed against it, and the header of
#: `speakers.yml` says so in words. Mirrored by `ANCHOR` in
#: `app/scripts/example-dates.mjs` and pinned to it by
#: `tools/tests/fixtures/example-dates.json`.
ANCHOR: Final = "2026-09-03"

#: The one environment variable that replaces the clock here. See the module
#: docstring for its single caller.
TODAY_ENV: Final = "CONVENER_EXAMPLE_TODAY"

#: An ISO day, bounded on both sides so that a longer run of digits is not
#: half-matched. Deliberately the whole of what this module recognises as a
#: date: `18:00` is not one, `MRG-5` is not one, and a four-digit year on its
#: own is not one either.
ISO_DAY: Final = re.compile(r"(?<!\d)(\d{4})-(\d{2})-(\d{2})(?!\d)")


def today(environ: Mapping[str, str] | None = None) -> str:
    """The day the example's records are read against.

    `CONVENER_EXAMPLE_TODAY` when it is set to an ISO day, and the real day
    otherwise. An unusable value is refused rather than ignored: a renderer
    that meant to pin the clock and misspelled the day would otherwise take a
    picture of a moving fixture and report success (D-25).
    """
    source = os.environ if environ is None else environ
    pinned = source.get(TODAY_ENV, "")
    if not pinned:
        return dt.date.today().isoformat()
    if _parse(pinned) is None:
        raise ValueError(
            f"{TODAY_ENV} is {pinned!r}, which is not a YYYY-MM-DD day -- "
            "it is read to pin the example instance's own records to one "
            "moment, and a value nothing can parse would leave them moving"
        )
    return pinned


def shift_days(day: str) -> int:
    """Whole days -- always a multiple of seven -- from the anchor to the week
    `day` falls in. Negative when `day` is before the anchor, which is the
    ordinary case for a pinned renderer."""
    now = _parse(day)
    if now is None:
        raise ValueError(f"{day!r} is not a YYYY-MM-DD day")
    anchor = _parse(ANCHOR)
    assert anchor is not None, "ANCHOR is not a day"
    return ((now - anchor).days // 7) * 7


def shifted(text: str, day: str) -> str:
    """`text` with every ISO day in it moved into the week of `day`.

    A date the calendar does not have is left exactly as it was found: this
    replaces text, and text it cannot read is text it has nothing to say
    about. The validator is what refuses such a value, by name, and it must
    go on seeing the value somebody actually typed.
    """
    days = shift_days(day)
    if days == 0:
        return text
    delta = dt.timedelta(days=days)

    def move(match: re.Match[str]) -> str:
        found = _parse(match.group(0))
        return match.group(0) if found is None else (found + delta).isoformat()

    return ISO_DAY.sub(move, text)


def _parse(day: str) -> dt.date | None:
    try:
        return dt.date.fromisoformat(day)
    except ValueError:
        return None
